src/librustc/ty/relate.rs

```diff
@@ -230,7 +230,7 @@ impl<'tcx> Relate<'tcx> for ty::ProjectionTy<'tcx> {
             Err(TypeError::ProjectionNameMismatched(
                 expected_found(relation, &a.item_name(tcx), &b.item_name(tcx))))
         } else {
-            let trait_ref = relation.relate(&a.trait_ref, &b.trait_ref)?;
+            let trait_ref = relation.relate(&a.trait_ref(tcx), &b.trait_ref(tcx))?;
```

Comment:
Relate substitutions directly. Also the name check above is silly now, should check `DefId`.